peps/pep-9999.rst

```diff
@@ -0,0 +1,322 @@
+
+This PEP does **not** propose the outright removal of support for ``Home-page``
+or ``Download-URL``. However, see :ref:`future-considerations` for
+thoughts on how a new (as of yet unspecified) major core metadata version
+could complete the deprecation cycle via removal of these deprecated fields.
+
+.. _package-indices:
+
+Package indices
+---------------
+
+This PEP stipulates the following for package indices:
+
+* When interpreting a distribution's metadata of version 1.2 or later
+  (e.g. for rendering on a web page), the index **SHOULD** prefer
+  ``Project-URL`` fields as a source of URLs over ``Home-page`` and
+  ``Download-URL``, even if the latter are explicitly provided.
+
+* If a distributions's metadata contains only the ``Home-page`` and
+  ``Download-URL`` fields, the index **MAY** choose to ignore those fields
+  and behave as though no URLs were provided in the metadata. In this case,
+  the index **SHOULD** present an appropriate warning or notification to
+  the uploading party.
+
+  * The mechanism for presenting this warning or notification is not
+    specified, since it will vary by index. By way of example, an index may
+    choose to present a warning in the HTTP response to an upload request, or
+    send an email or other notification to the maintainer(s) of the project.
+
+* If a distribution's metadata contains both sets of fields, the index **MAY**
+  choose to reject the distribution outright. However, this is
+  **NOT RECOMMENDED** until a future unspecified major metadata version
+  formally removes support for ``Home-page`` and ``Download-URL``.
+
+
+* Any changes to the interpretation of metadata of version 1.2 or later that
+  result in previously recognized URLs no longer being recognized
+  **SHOULD NOT** be retroactively applied to previously uploaded packages.
+
+These stipulations do not change the optionality of URL processing by indices.
+In other words, an index that does not process URLs within uploaded
+distributions may continue to ignore all URL fields entirely.
+
+.. _conventions-for-labels:
+
+Conventions for ``Project-URL`` labels
+======================================
+
+The deprecations proposed above require a formalization of the currently
+informal relationship between ``Home-page``, ``Download-URL``, and their
+``Project-URL`` equivalents.
+
+This formalization has two parts:
+
+1. A set of rules for canonicalizing ``Project-URL`` labels;
+2. A set of "well-known" canonical label values that indices may specialize
+   URL presentation for.
+
+Label canonicalization
+----------------------
+
+The core metadata specification stipulates that ``Project-URL`` labels are
+free text, limited to 32 characters.
+
+This PEP proposes adding the concept of a "canonicalized" label to the core
+metadata specification. Label canonicalization is defined via the following
+Python function:
+
+.. code-block:: python
+
+    import string
+    def canonicalize_label(label: str) -> str:
```

Comment:
Consider also using Unicode normalization, such as NFC.

Reply:
No objection, although this has no impact on the current all-ASCII well-known labels -- any concerns @ncoghlan?